### PR TITLE
Add new checker `consider-using-tuple`

### PR DESCRIPTION
**Pull request by cdce8p**
## Description
Detect when iterating over an in-place `list` or `set` in a for-loop or comprehension. Suggest to use a `tuple` instead.

```py
# old
for x in [1, 2, 3]: pass
[x for x in [1, 2]]

# new
for x in (1, 2, 3): pass
[x for x in (1, 2)]
```

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
What is the rational ? Are performance better because a tuple is immutable ?

**Comment by cdce8p** [comment]
> What is the rational ? Are performance better because a tuple is immutable ?

It should be, also probably not by much. Depends on the situation. I would consider it similar to the way we suggest generator expressions instead of lists. For small lists the benefit is also negligible.

Do you think this would make `pylint` too opinionated? I debated that myself, especially together with the `consider-using-namedtuple` checker I added a few days ago. In the end they are only suggestions and technically also lead to better code. I'm interested in what you think.

**Comment by cdce8p** [comment]
A possible addition for another MR if this one is accepted, could be:
```py
var = [1, 2]

# old
var.extend([1, 2])

# new
var.extend((1, 2))
```

In that case it would make sense to change the msg symbol to just `consider-using-tuple`.

**Comment by Pierre-Sassoulas** [comment]
> Do you think this would make pylint too opinionated? I debated that myself, especially together with the consider-using-namedtuple checker I added a few days ago. 

Ideally I would like to be able to explain very clearly what is better in the code if you follow the suggestion made by pylint by default.  Basic test for a default checker: would a a senior dev know and explain for sure what it the benefit of the proposed approach ? (For example for argument-differs I expect a senior dev to come easily with : "You broke the L in Solid, that's bad because ..."). And I'd like the time it took to make the change and the time it took to calculate the warning to be worth it without a doubt. The small benefit, slow check or opinionated checks should be optional. For example ``consider-using-a-generator`` is great, but I would not bet the generator is better, it should probably be optional.

I think we have two problems:
- pylint is becoming slower and slower (Nowadays it's often taking more than 10s to run pylint in pre-commit for me, and that's a limit in term of keeping the user attention. We're also going to start annoying IDE users with really powerful PC that can't run pylint on save reasonably fast.). 
- We also have a bunch of warning that are not worth their calculation time or actively harmful because often it's a false positive or too opinionated (see #3512). It makes pylint less useful because the quality of warnings is low, noise is high and everything get disabled as a result - even the good warnings.

For example of "empty-docstring", (that I accepted) is a default check but it's not bringing a lot of value. empty docstrings could be removed automatically by an opinionated auto-formatter. But forcing human to do it with a warning is not a good use of developer time (especially when we saw that teachers are using the tool to mark student blindly (!), we're going to make potential dev change career with those pesky warnings).

In this particular MR case, it seems - although I could not be sure of the reason myself - that there is a small benefit for this checker, so it could become an optional plugin.

But going forward we need a way to measure performance and what time is a checker costing and also remove the checks that are annoying and not very important from the default checks in order to cut down runtime.  Also, possibly not taking time to calculate what is already done by flake8 faster than us (without inference).

@AWhetter made a recap of that and it probably took a lot of time in #3512 , feel free to check and/or upgrade with new checkers we added since.

**Comment by cdce8p** [comment]
@Pierre-Sassoulas Thanks for that detailed explanation!

**TLDR:** The main benefits I think `pylint` provides over all other checkers, are the checks for code consistency that are by their nature more (time) expensive and sometimes opinionated. Moving them to an optional plugin / or removing them entirely would probably mean removing the reason to use `pylint` in the first place.

Therefore, it would make sense to add `consider-using-tuple` as a default check.

---

> pylint is becoming slower and slower (Nowadays it's often taking more than 10s to run pylint in pre-commit for me, and that's a limit in term of keeping the user attention. We're also going to start annoying IDE users with really powerful PC that can't run pylint on save reasonably fast.).

I'm seeing that as well, although I'm not sure pylint should be a pre-commit hook in the first place. From my point of view, it is much more useful in a CI stage. Consider this, the easy problems are already found by the likes of `flake`, `pyupgrade`, `pylance` / `pyright`, `mypy`, and `black`. Often times they are also much faster as well. The main benefits I see with `pylint` are the more complex checks that all the other ones don't provide or don't want to implement. These are inherently more time consuming and sometimes opinionated, but they make sure the code follows a consistent, hopefully better, style.

`consider-using-a-generator`, `consider-using-namedtuple`, and also this one, `consider-using-tuple`, are great examples for that. It isn't wrong to write `[x for x in [1, 2]]` but for large in-place lists `[x for x in (1, 2)]` is a bit more efficient as it's immutable.

You mentioned `empty-docstring`. I would consider it somewhat similar. The point isn't for an auto-formatter to remove it, it's to make sure a code base uses a consistent style. It's not really necessary to know that during writing and debugging, but once the code gets submitted in a MR it's quite useful to have `pylint` check it to guarantee consistency.

Every new checker we implement reduces the need for reviewers to notice code inconsistencies during reviews. I wouldn't be surprised if some bigger corporations have many additional custom checkers. (Although I hope they would contribute them back to us.)

> We also have a bunch of warning that are not worth their calculation time or actively harmful because often it's a false positive or too opinionated (see #3512). It makes pylint less useful because the quality of warnings is low, noise is high and everything get disabled as a result - even the good warnings.

I'm not sure that is really the case. Yeah, there are a few that I disable by default for my personal projects (e.g., `too-many-*`, `too-few-*`, `missing-docstring`) but aside from that I don't need to add any more `disables` then I do for `mypy`.  Once you know the style it's quite easy to write the code. Having the CI checks just makes sure I don't miss anything.

> But forcing human to do it with a warning is not a good use of developer time

I disagree. If we don't mention the errors, the senior dev doing the review has to do it which would be worse. Having the automated feedback early helps the dev as they can fix it before anybody else takes a look at it. It also helps the one doing the review as he/she can now focus on the logic itself and doesn't need to make style comments.

> (especially when we saw that teachers are using the tool to mark student blindly (!), we're going to make potential dev change career with those pesky warnings).

That was truly unfortunate. Those teachers haven't understood how to use these tools properly. If they would, they would have mentioned to their students to check the code themselves before they submit it. That's how I remember it from my programming course. Only if we didn't fix it **after** we knew the error, was it a reason for a lower score.

--
> In this particular MR case, it seems - although I could not be sure of the reason myself - that there is a small benefit for this checker, so it could become an optional plugin.

I though about that again since yesterday. It would be easy enough to move this check to an optional plugin. If we do that, probably many others can be moved there as well. However, what would then be the true benefit `pylint` would provide over all the others? Without these code consistency checks there isn't much that the others don't do. Just take a look at `invalid-all-format` that I added yesterday: `pylance` would have also caught that. (That alone isn't a reason not implement it in pylint though.)

To further prove my point, before I started working on the MR, I suggested it for `pyupgrade` only to be rejected a few minutes after I opened the issue: https://github.com/asottile/pyupgrade/issues/458

In the end, I believe it would make sense to keep it as a refactoring checker without making it optional.

**Comment by Pierre-Sassoulas** [comment]
Sorry for the short answer, I don't have much time. It is true that having a lot of check is what is distinguishing pylint and we should not remove too much. But there's a base quality of warning to take into account, having a lot of questionable warning is not a good thing. Also the performance/value ratio need to be considered. Gamification is also a distinguishing characteristic of pylint. There was an (imho) very good idea by PcManticore that we could have multiple stages, the tough and opinionated one being shown only when the major and obvious one are fixed you "level up" and get the more opinionated fix (I could try to find the comment, if you're interested). 

Of course this is a much larger discussion than the ``consider-using-tuple`` checker, I'm not saying this particular checker is not worthy. 

Random ideas:
- Keep 80/90% of current checkers in default and removing the more questionable one
- Adding a ``--non-opinionated`` option that would keep ``use-a-generator`` but not ``consider-using-generator``.
- Having an option that launch all the optional checker even if they are not loaded in the pylintrc.
- Include a well written rational to explain the reason behind a check to be more friendly toward users.
- Include an "opinionated" boolean in the Message Definition

**Comment by cdce8p** [comment]
> Sorry for the short answer, I don't have much time.

No problem! My reply was probably too long anyway 😅

> Gamification is also a distinguishing characteristic of pylint.

I haven't considered that as I'm more used to all or nothing, but you're absolutely right that is a good point.

> There was an (imho) very good idea by PcManticore that we could have multiple stages, the tough and opinionated one being shown only when the major and obvious one are fixed you "level up" and get the more opinionated fix

I'm not sure this would be helpful, at least not in a CI context. If you consider that runs can sometimes take 10-20min, having to rerun it just to discover that you need to fix something you didn't know about before could be quite frustrating, I imaging.

> `--non-opinionated` option and
> Include an "opinionated" boolean in the Message Definition

I like that! That could be something to work on after the release of `2.9`

--
How do we want to proceed with this MR then? I still think this checker shouldn't be put in an extension, but it obviously falls in the `opinionated` category. Is it possible to disable checkers in the code (not through pylintrc) without removing them? That way we could still include it, but it would only be activated if it's explicitly enabled.

If that's possible, it might be something we could consider for `consider-using-namedtuple-or-dataclass` as well.

**Comment by Pierre-Sassoulas** [comment]
>  Is it possible to disable checkers in the code (not through pylintrc) without removing them? That way we could still include it, but it would only be activated if it's explicitly enabled.

Yes that's what I meant when I'm saying non builtin checkers. We can activate the non default plugin with ``load_plugin`` in pylintrc (for example confusing_elif is a non builtin checker, there's a list : http://pylint.pycqa.org/en/latest/technical_reference/extensions.html). I think we should provide an easy way to add all the plugin checkers and modify the pylintrc easily. Right now if a checker is not a default one I think there's not a lot of chance it's used. But we could make it easier. We're also not forced to add indefinitely to the existing default checker, we can also add new clean checkers to the default one.

> I haven't considered that as I'm more used to all or nothing, but you're absolutely right that is a good point.

Never thought of that either there's only the score that approach the concept of gamification right now. And some user sees that as simply annoying and rude (Score -56/10 == "your code sux"). But seeing PCManticore talk about it, it really made sense.  Gamification is powerful and we can expand it beyond the current score.

I thought a little about gamification and making pylint more friendly; and I think we can keep the CI pylintrc like it is right now but we can have a ``pylint-guide`` command or ``pylint --guide`` command that would keep the noise to a minimum and actually make pylint more friendly at first and also after that with level up. Showing only the 5 most pressing issues in the repo at a time and upgrading the pylintrc step by step. The experience I have right now in starting to use pylint on a code base is overwhelming, I have to use  ``pylint .|grep -o "([a-z\-]*)$"|cut -d "(" -f2|sort|uniq|sed s/\)/,/`` to disable all the existing warning and start from somewhere. Then I need to think hard before enabling them again to know where to start. The way I see it the pylint guide would add the disable automatically, maybe in pylintrc, maybe in code directly, then it would rank the warning and tell you exactly what to do next and provide a feeling of progression with level up and "encouragements". ``Congratulation you do not have any error level warning``, ``Congratulation the code follow SOLID principles``, ``Congratulation we do not have any performance improvement to provide``, ``Congratulation your code did not regress after upgrading to pylint 2.10.1, you were already that good``, ``Congratulation you've got a perfect code according to pylint's opinion`` etc. A lot of variables to follow that could be in the pylintrc.

I'm going to create an issue for this as this MR is not the place.

**Comment by cdce8p** [comment]
I've move the checker to a newly created `CodeStyleChecker` extension.

File: doc/whatsnew/2.9.rst [hunk 1]
```diff
@@ -39,6 +39,11 @@ New checkers
 
 * New checker ``await-outside-async``: Emitted when await is used outside an async function.
 
+* Add new extension ``CodeStyleChecker``. It includes checkers that can improve code
+  consistency. As such they don't necessarily provide a performance benefit
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/2.9.rst:43]: But this one provide a small performance benefit right ?

**Review by Pierre-Sassoulas — APPROVED** [review]
This new checker name and description is nice, I think we'll use it a lot for opinionated checks. But maybe this check in particular need yet another one named 'smallPerformanceImprovment' ?  What donc you think ?

File: doc/whatsnew/2.9.rst (hunk 1 above)
Comment (cdce8p) [reply doc/whatsnew/2.9.rst:43]: Yeah it does, although it's probably a negligible aside from some edge cases.

I don't think it would make sense to create yet another extension just for such cases. We should think about this more in  these categories:
1. Does a checker provide a clear benefit, i.e. detect a common issue or improve performance => it should probably be part of the core checker classes
2. Is it something that would improve code consistency, maybe because it's slightly better with regards to performance and therefore preferred => the new extension is probably the right place
3. Everything else should go into another extension

The `CodeStyleChecker` extension shouldn't be the catch all one for everything that isn't good enough for the core. Especially if we want devs to actually use it.

Comment (Pierre-Sassoulas) [reply doc/whatsnew/2.9.rst:43]: > The CodeStyleChecker extension shouldn't be the catch all one for everything that isn't good enough for the core. Especially if we want devs to actually use it.

Makes perfect sense !
